### PR TITLE
fix #4280: only drop points matching WHERE clause

### LINKED ISSUE #4280: [0.9.4.1] DROP SERIES with a field in the WHERE clause always drops all points in the measurement
I have got a series with values equal to 0. Since we decided not to push any value that is 0 in our InfluxDB, I want to clean up this serie and remove the existing values equal to 0. Here is what I did:

```
> select count(value) from p2p
name: p2p
---------
time      count
1970-01-01T00:00:00Z  286216

> select count(value) from p2p where value=0
name: p2p
---------
time      count
1970-01-01T00:00:00Z  113630

> drop series from p2p where value=0
> select count(value) from p2p where value=0
> select count(value) from p2p
name: p2p
---------
time      count
1970-01-01T00:00:00Z  3
```
I am surprised that is removed everything (the 3 entries is because the DB get populated at the same time by live clients).

Is the DROP series for a specific value supported?

### PR DESCRIPTION
**Pull request by dgnorton**

### CODE REVIEW

**Comment by otoolep** [comment]
I'm confused -- I only see tests that check if certain errors occur. Is there a bug here that was fixed? did those errors not get spit out before, and instead data was dropped?

File: influxql/ast.go [hunk 1]
```diff
@@ -2698,6 +2653,47 @@ func CloneExpr(expr Expr) Expr {
 	panic("unreachable")
 }
 
+// HasTimeExpr returns true if the expression has a time term.
+func HasTimeExpr(expr Expr) bool {
```
Comment (otoolep) [inline influxql/ast.go]: Why this refactor?

Comment (dgnorton) [inline influxql/ast.go]: Before it was a method on `type SelectStatement` but really it should be a function that works on an `Expr`.  Some of the meta-queries now use it also.

File: cmd/influxd/run/server_test.go [hunk 2]
```diff
@@ -382,6 +382,24 @@ func TestServer_Query_DropSeriesFromRegex(t *testing.T) {
...
 			params:  url.Values{"db": []string{"db0"}},
 		},
+		&Query{
+			name:    "Drop series with WHERE field should error",
+			command: `DROP SERIES FROM c WHERE val > 50.0`,
+			exp:     `{"results":[{"error":"DROP SERIES doesn't support fields in WHERE clause"}]}`,
+			params:  url.Values{"db": []string{"db0"}},
+		},
+		&Query{
+			name:    "make sure DROP SERIES with field in WHERE didn't delete data",
```
Comment (dgnorton) [inline cmd/influxd/run/server_test.go]: @otoolep this is the new test that makes sure the original issue is fixed.

**Comment by dgnorton** [comment]
@otoolep  yes, issue #4280 was for data that was getting dropped when it shouldn't.  While fixing that I also went through the other queries making sure they also check for things in the `WHERE` clause that should return errors.

File: tsdb/meta.go [hunk 3]
```diff
@@ -833,10 +833,30 @@ func (m *Measurement) idsForExpr(n *influxql.BinaryExpr) (SeriesIDs, influxql.Ex
...
 }
 
+// FilterExprs represents a map of series IDs to filter expressions.
+type FilterExprs map[uint64]influxql.Expr
+
+// DeleteBoolLiteralTrues deletes all elements whose filter expression is a boolean literal true.
+func (fe FilterExprs) DeleteBoolLiteralTrues() {
+	for id, expr := range fe {
+		if e, ok := expr.(*influxql.BooleanLiteral); ok && e.Val == true {
+			delete(fe, id)
```
Comment (otoolep) [inline tsdb/meta.go]: Is it safe to be modifying a map while iterating it? Don't you need to build the list of keys, and then loop over that list?

Comment (dgnorton) [inline tsdb/meta.go]: It's safe.  From the golang spec...
```
The iteration order over maps is not specified and is not guaranteed to be the same from one
iteration to the next. If map entries that have not yet been reached are removed during iteration,
the corresponding iteration values will not be produced. If map entries are created during
iteration, that entry may be produced during the iteration or may be skipped.
```
https://golang.org/ref/spec#For_statements

Comment (otoolep) [inline tsdb/meta.go]: Nice, this will allow me to simplify some code I wrote recently.

**Comment by otoolep** [comment]
So do we need a test showing that only certain series are now deleted when there is a WHERE clause? Perhaps I am dense and missing it.

**Comment by otoolep** [comment]
Oh, I see, `DROP SERIES` with `WHERE` should be blocked and just don't allow it?

**Comment by dgnorton** [comment]
@otoolep no, `DROP SERIES FROM cpu WHERE tagKey = 'tagVal'` is allowed but `WHERE time` and `WHERE value` aren't currently supported.

**Comment by otoolep** [comment]
Had a quick review over hangout, and I follow this code. +1

**Pull request merged**